packages/flutter/lib/src/widgets/scrollable.dart

```diff
@@ -763,3 +868,121 @@ class _RenderScrollSemantics extends RenderProxyBox {
+    this.type = ScrollIncrementType.line,
+  })  : assert(axis != null),
+        assert(type != null),
+        super(ScrollAction.key);
+
+  /// The direction in which to scroll the scrollable containing the focused
+  /// widget.
+  final Axis axis;
+
+  /// Whether or not the natural scroll direction for this scroll operation
+  /// should be reversed.
+  final bool reversed;
+
+  /// The type of scrolling that is intended.
+  final ScrollIncrementType type;
+
+  @override
+  bool isEnabled(BuildContext context) {
+    return Scrollable.of(context) != null;
+  }
+}
+
+/// An [Action] that scrolls the [Scrollable] that encloses the current
+/// [primaryFocus] by the amount configured in the [ScrollIntent] given to it.
+class ScrollAction extends Action {
+  /// Creates a const [ScrollAction].
+  ScrollAction() : super(key);
+
+  /// The [LocalKey] that uniquely connects this action to a [ScrollIntent].
+  static const LocalKey key = ValueKey<Type>(ScrollAction);
+
+  // Find out how much of an increment to move by, taking the different
+  // directions into account.
+  double _getIncrement(ScrollableState state, ScrollIntent intent) {
+    final double increment = state.calculateIncrement(type: intent.type);
+    final AxisDirection resolvedDirection = getAxisDirectionFromAxisReverseAndDirectionality(state.context, intent.axis, intent.reversed);
+    switch (resolvedDirection) {
+      case AxisDirection.down:
+        switch (state.axisDirection) {
+          case AxisDirection.up:
+            return -increment;
+            break;
+          case AxisDirection.down:
+            return increment;
+            break;
+          case AxisDirection.right:
+          case AxisDirection.left:
+            return 0.0;
+        }
+        break;
+      case AxisDirection.up:
+        switch (state.axisDirection) {
+          case AxisDirection.up:
+            return increment;
+            break;
+          case AxisDirection.down:
+            return -increment;
+            break;
+          case AxisDirection.right:
+          case AxisDirection.left:
+            return 0.0;
+        }
+        break;
+      case AxisDirection.left:
+        switch (state.axisDirection) {
+          case AxisDirection.right:
+            return -increment;
+            break;
+          case AxisDirection.left:
+            return increment;
+            break;
+          case AxisDirection.up:
+          case AxisDirection.down:
+            return 0.0;
+        }
+        break;
+      case AxisDirection.right:
+        switch (state.axisDirection) {
+          case AxisDirection.right:
+            return increment;
+            break;
+          case AxisDirection.left:
+            return -increment;
+            break;
+          case AxisDirection.up:
+          case AxisDirection.down:
+            return 0.0;
+        }
+        break;
+    }
+    return 0.0;
+  }
+
+  @override
+  void invoke(FocusNode node, ScrollIntent intent) {
```

Comment:
Can this method ever be called before the first layout has happened?  If so, we're accessing values here that might be null, e.g. if we're in the first build of the widget and we're trying to access the `state.position.pixels`, as well as the `state.position.viewportDimensions` (for the default calculateIncrement impl).

Reply:
It would be unlikely (a key event would have to arrive and be handled before the first frame), but better safe than sorry. Added null checks.

Reply:
Actually, this is probably impossible: initial focus would have to also be set, and the earliest that can happen is the second frame.

Reply:
Maybe just some asserts then?

Reply:
Done.